cdap/resource_streaming_program_run.go

```diff
@@ -0,0 +1,325 @@
+
+	argsObj := make(map[string]string)
+	// cast map[string]interface{} to map[string]string
+	for k, val := range d.Get("runtime_arguments").(map[string]interface{}) {
+		argsObj[k] = val.(string)
+	}
+
+	runID, err := uuid.NewRandom()
+	if err != nil {
+		return fmt.Errorf("error generating uuid for faux run id: %v", err)
+	}
+	// This runtime arg will be unused by the pipeline but will allow the provider to associate a run with this resource.
+	argsObj[fauxRunID] = runID.String()
+
+	b, err := json.Marshal(argsObj)
+	if err != nil {
+		return err
+	}
+
+	req, err := http.NewRequest(http.MethodPost, startAddr, bytes.NewReader(b))
+	if err != nil {
+		return err
+	}
+
+	if _, err := httpCall(config.httpClient, req); err != nil {
+		return err
+	}
+
+	// Poll until actually reaches RUNNING state.
+	return resource.Retry(d.Timeout(schema.TimeoutCreate), func() *resource.RetryError {
+		time.Sleep(10 * time.Second) // avoid spamming retries and initial failure to find run.
+		isRunning, err := isFauxRunIDRunningYet(config, runsAddr, runID.String())
+		if err != nil {
+			return resource.NonRetryableError(err)
+		}
+		if isRunning {
+			d.SetId(runID.String())
+			return nil
+		}
+		return resource.RetryableError(fmt.Errorf("still waiting for program run with faux run id: %v which is in an initializing state", runID.String()))
+	})
+}
+
+//TODO ?
+func resourceStreamingProgramRunRead(d *schema.ResourceData, m interface{}) error {
+	return nil
+}
+
+type runtimeArgs struct {
+	FauxRunID string `json:"__FAUX_RUN_ID__"`
+}
+
+type runtimeProperties struct {
+	RuntimeArgs *runtimeArgs `json:"runtimeArgs"`
+}
+
+func (ra *runtimeArgs) UnmarshalJSON(data []byte) error {
+	unquoted, err := strconv.Unquote(string(data))
+	if err != nil {
+		return fmt.Errorf("failed to escape runtime arguments %v: %v", string(data), err)
+	}
+
+	// Use alias to avoid infinite recursion: https://stackoverflow.com/q/52433467
+	type alias runtimeArgs
+	var a alias
+	if err := json.Unmarshal([]byte(unquoted), &a); err != nil {
+		return err
+	}
+	*ra = runtimeArgs(a)
+	return nil
+}
+
+// These are the only keys we need
+type run struct {
+	RunID      string            `json:"runid"`
+	Status     string            `json:"status"`
+	Properties runtimeProperties `json:"properties"`
+}
+
+// Checks if there is a running run for the terraform faux run id
+// raises error if the program is not in an initializing state (e.g. it failed or was killed in the ui)
+func isFauxRunIDRunningYet(config *Config, runsAddr string, runID string) (bool, error) {
+	r, err := getRunByFauxID(config, runsAddr, runID)
+	if err != nil {
+		return false, err
+	}
+
+	if r.Status == "RUNNING" {
+		return true, nil
+	}
+	if !programRunInitializingStatuses[r.Status] {
+		return false, fmt.Errorf("program not running or initializing, in state: %v", r.Status)
+	}
+	return false, nil
+
+}
+
+// TODO optimization: we call this function often when we could probably get the real runid once and cache it.
```

Comment:
FYI the right way would probably be when creating the run to get the run id and store as a computed field.